ddht/v5_1/alexandria/network.py

```diff
@@ -0,0 +1,131 @@
+import logging
+from typing import AsyncContextManager, Optional, Type
+
+from async_service import Service
+from eth_enr import ENRManagerAPI
+from eth_typing import NodeID
+import trio
+
+from ddht.base_message import InboundMessage
+from ddht.endpoint import Endpoint
+from ddht.exceptions import DecodingError
+from ddht.subscription_manager import SubscriptionManager
+from ddht.v5_1.abc import NetworkAPI
+from ddht.v5_1.alexandria.abc import AlexandriaClientAPI, AlexandriaNetworkAPI
+from ddht.v5_1.alexandria.constants import ALEXANDRIA_PROTOCOL_ID
+from ddht.v5_1.alexandria.messages import (
+    PingMessage,
+    TAlexandriaMessage,
+    decode_message,
+)
+from ddht.v5_1.alexandria.payloads import PongPayload
+from ddht.v5_1.messages import TalkRequestMessage, TalkResponseMessage
+
+
+class AlexandriaNetwork(Service, AlexandriaNetworkAPI):
+    logger = logging.getLogger("ddht.Alexandria")
+
+    protocol_id = ALEXANDRIA_PROTOCOL_ID
+
+    _alexandria_client: AlexandriaClientAPI
+
+    def __init__(self, client: AlexandriaClientAPI) -> None:
+        self._alexandria_client = client
+        self.protocol_id = self._alexandria_client.protocol_id
+
+        self.subscription_manager = SubscriptionManager()
+
+    @property
+    def network(self) -> NetworkAPI:
+        return self._alexandria_client.network
+
+    @property
+    def enr_manager(self) -> ENRManagerAPI:
+        return self._alexandria_client.network.enr_manager
+
+    def subscribe(
+        self,
+        message_type: Type[TAlexandriaMessage],
+        endpoint: Optional[Endpoint] = None,
+        node_id: Optional[NodeID] = None,
+    ) -> AsyncContextManager[
+        trio.abc.ReceiveChannel[InboundMessage[TAlexandriaMessage]]
+    ]:
+        return self.subscription_manager.subscribe(message_type, endpoint, node_id)  # type: ignore
+
+    async def run(self) -> None:
+        self.manager.run_daemon_task(self._feed_talk_requests)
+        self.manager.run_daemon_task(self._feed_talk_responses)
+        self.manager.run_daemon_task(self._pong_when_pinged)
+
+        await self.manager.wait_finished()
+
+    #
+    # High Level API
+    #
+    async def ping(
+        self, node_id: NodeID, *, endpoint: Optional[Endpoint] = None,
+    ) -> PongPayload:
+        response = await self._alexandria_client.ping(node_id, endpoint=endpoint)
+        return response.payload
+
+    #
+    # Long Running Processes
+    #
+    async def _pong_when_pinged(self) -> None:
+        async with self.subscribe(PingMessage) as subscription:
+            async for request in subscription:
+                await self._alexandria_client.send_pong(
+                    request.sender_node_id,
+                    request.sender_endpoint,
+                    enr_seq=self.enr_manager.enr.sequence_number,
+                    request_id=request.request_id,
+                )
+
+    async def _feed_talk_requests(self) -> None:
+        async with self.network.client.dispatcher.subscribe(
```

Comment:
I like your reasoning on the other PR, I think you said that `network.client` present a unified interface. This looks like you're digging into internals but really the class has namespaced its methods for you for your convenience. That sounds great to me!

This line gives me more pause though :sweat_smile: This expands out into `self._alexandria_client.network.client.dispatcher`, which really feels like it's digging into implementation details. When any of these members are renamed or moved around the commit will need to also make unrelated changes: updating this class.

It's not a large harm, and I can't think of an easy way to fix it so if you also can't think of one I'm okay with merging this. Going forward I bet this could be simplified with the introduction of a message bus. Instead of needing to go and find the class which will let us subscribe to our messages we could listen for our messages.

Reply:
I'm going to mirror the subscribe API onto the `ClientAPI` which turns this into `alexandria_client.network.client.subscribe` which is still kind of *deep* into the object hierarchy, but it does eliminate the `DispatcherAPI` from the chain and further consolodates the *external* APIs onto the `ClientAPI` which is much more external facing than the dispatcher.